pwndbg/aglib/proc.py

```diff
@@ -0,0 +1,186 @@
+"""
+Provides values which would be available from /proc which
+are not fulfilled by other modules and some process/gdb flow
+related information.
+"""
+
+from __future__ import annotations
+
+import functools
+import sys
+from types import ModuleType
+from typing import Callable
+from typing import List
+from typing import Optional
+from typing import Tuple
+from typing import TypeVar
+
+from elftools.elf.relocation import Relocation
+from typing_extensions import ParamSpec
+
+import pwndbg.aglib.qemu
+import pwndbg.lib.cache
+import pwndbg.lib.memory
+
+P = ParamSpec("P")
+T = TypeVar("T")
+
+pid: int
+tid: int
+thread_id: int
+alive: bool
+thread_is_stopped: bool
+stopped_with_signal: bool
+exe: str | None
+binary_base_addr: int
+binary_vmmap: Tuple[pwndbg.lib.memory.Page, ...]
+# dump_elf_data_section: Tuple[int, int, bytes] | None
+# dump_relocations_by_section_name: Tuple[Relocation, ...] | None
+# get_section_address_by_name: Callable[[str], int]
+
+
+def OnlyWhenRunning(func: Callable[P, T]) -> Callable[P, T | None]: ...
+def OnlyWhenQemuKernel(func: Callable[P, T]) -> Callable[P, T]: ...
+def OnlyWithArch(
+    arch_names: List[str],
+) -> Callable[[Callable[..., T]], Callable[..., Optional[T]]]: ...
+
+
+class module(ModuleType):
+    @property
+    def pid(self) -> int:
+        # QEMU usermode emulation always returns 42000 for some reason.
+        # In any case, we can't use the info.
+        if pwndbg.aglib.qemu.is_qemu_usermode():
+            return pwndbg.aglib.qemu.pid()
+        return pwndbg.dbg.selected_inferior().pid()
+
+    @property
+    def tid(self) -> int:
+        if pwndbg.aglib.qemu.is_qemu_usermode():
+            return pwndbg.aglib.qemu.pid()
+        return pwndbg.dbg.selected_thread().ptid()
+
+    @property
+    def thread_id(self) -> int:
+        return pwndbg.dbg.selected_thread().index()
+
+    @property
+    def alive(self) -> bool:
+        """
+        Informs whether the process has a thread. However, note that it will
+        still return True for a segfaulted thread. To detect that, consider
+        using the `stopped_with_signal` method.
+        """
+        return pwndbg.dbg.selected_inferior().alive()
+
+    @property
+    def exe(self) -> str | None:
+        """
+        Returns the debugged file name.
+
+        On remote targets, this may be prefixed with "target:" string.
+        See this by executing those in two terminals:
+        1. gdbserver 127.0.0.1:1234 /bin/ls
+        2. gdb -ex "target remote :1234" -ex "pi pwndbg.gdblib.proc.exe"
+
+        If you need to process the debugged file use:
+            `pwndbg.gdblib.file.get_proc_exe_file()`
+            (This will call `pwndbg.gdblib.file.get_file(pwndbg.gdblib.proc.exe, try_local_path=True)`)
+        """
+
+        return pwndbg.dbg.selected_inferior().main_module_name()
+
+    @property
+    @pwndbg.lib.cache.cache_until("start", "stop")
+    def binary_base_addr(self) -> int:
+        return self.binary_vmmap[0].start
```

Comment:
Should we return here `None` if index `0` not exists?

Reply:
Well, while I think it would make sense to do that from general API design standpoint, I don't think it's a great idea to do that in this PR specifically.

Here's the thing. That piece of code was lifted straight off `pwndbg/gdblib/proc.py`, so that's a potential bug that we share with `gdblib`, meaning that all code that currently exists in Pwndbg, including the one being ported to `aglib` in this PR, will probably not behave gracefully at all if we returned `None` there for that case, anyway.

I agree that we should fix that, but having the return `None` case would kick the can of handling the missing vmmap down to every part of Pwndbg that calls into that function, and so I don't think this PR is the place to do it.